Edit-message

asv-tungpt opened the pull request.

asv-tungpt committed:
Edit-message

phucnpm commented on djangoguestbook/guestbook/templates/guestbook/main_page.html:
```diff
@@ -10,6 +10,24 @@
+					<div><input type="submit" value="Submit"></div>
+				</form>
+			{% else %}
+		   		<form action="/edit/" method="edit">
```
method="edit" ???
- hãy dùng PUT

phucnpm commented on djangoguestbook/guestbook/templates/guestbook/main_page.html:
```diff
@@ -10,6 +10,24 @@
+				<blockquote>Updated by:  {{ greeting.updateby|escape}}</blockquote>
+			<blockquote>Updated date:  {{ greeting.updatedate|escape}}</blockquote>
+			{% endif %}
+			{% if edited_id == greeting.key.id %}
```
điều kiện này nghĩa là gì ? 
yêu cầu là nếu là administrator or owner thì được phép edit message
- làm đơn giản
  - hiển thị button edit
  - sau khi click thì mở sang View mới để edit
  - edit thành công thì redirect về trang chủ

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -5,6 +5,46 @@
+		self.__class__.edited_id = long(self.request.GET.get('edited_message'))
+		self.__class__.old_guestbook_name = self.request.GET.get('old_guestbook_name')
+		guestbook_name = self.request.GET.get('guestbook_name', DEFAULT_GUESTBOOK_NAME)
+		greetings = Greeting.get_lastest(guestbook_name, 10, False)
```
không handle việc hiển thị nội dung trong View này, hãy để cho GreetingView làm

asv-tungpt committed:
edit-message

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -44,6 +47,16 @@ def put_from_dict(cls, dict):
 		greeting.put()
 
+	@classmethod
+	def edit_greeting(cls, dict):
```
hãy đổi `edit_greeting` thành `update_from_dict`

asv-tungpt committed:
Change name method and the way return

- edit_greeting to update_from_dict
- return to default method from_valid instead of HttpResponseRedirect

phucnpm commented on djangoguestbook/guestbook/templates/guestbook/main_page.html:
```diff
@@ -10,6 +10,31 @@
+						<div><input type="hidden" value={{ guestbook_name }} name ="guestbook_name"></div>
+					</form>
+				{% else %}
+					<form action="/edit/" method="put">
```
hãy move edit form vào template của EditGreetingView
- Khi bấm button edit -> mở trang EditGreetingView -> fill giá trị cũ của greeting (content, guestbook) vào edit form

asv-tungpt committed:
create new template for EditGreetingView

asv-tungpt committed:
Bỏ những dòng

phucnpm commented on djangoguestbook/guestbook/templates/guestbook/main_page.html:
```diff
@@ -10,6 +10,20 @@
+				<form action="/edit/" method="put">
+					<blockquote></blockquote>
+					<div><input type="submit" value="Edit message"></div>
+					<div><input type="hidden" value={{ greeting.key.id|escape }} name ="edited_message"></div>
```
đổi `edited_message` thành `greeting_id`

phucnpm commented on djangoguestbook/guestbook/templates/guestbook/main_page.html:
```diff
@@ -10,6 +10,20 @@
+					<blockquote></blockquote>
+					<div><input type="submit" value="Edit message"></div>
+					<div><input type="hidden" value={{ greeting.key.id|escape }} name ="edited_message"></div>
+					<div><input type="hidden" value={{ guestbook_name }} name ="old_guestbook_name"></div>
```
`old_guestbook_name` sẽ xài ở đâu ?
hãy sửa thành `guestbook_name`

phucnpm commented on djangoguestbook/guestbook/urls.py:
```diff
@@ -1,8 +1,10 @@
 	url(r'^$', GreetingView.as_view(), name='home'),
 	url(r'^sign/$', GreetingView.as_view(), name='sign'),
+	url(r'^edit/$', EditGreetingView.as_view(), name='edit'),
+	url(r'^edited/$', EditGreetingView.as_view(), name='edited'),
```
đã tách edit form ra khỏi template của GreetingView thì không cần define 
`url(r'^edited/$', EditGreetingView.as_view(), name='edited'),`

phucnpm commented on djangoguestbook/guestbook/templates/guestbook/edit_page.html:
```diff
@@ -0,0 +1,16 @@
+	<title></title>
+</head>
+<body>
+	<form action="/edited/" method="post">
```
hãy tạo form từ EditForm với đủ các thành phần greeting_id, guestbook_name, content

phucnpm commented on djangoguestbook/guestbook/forms.py:
```diff
@@ -4,3 +4,7 @@
+
+
+class EditForm(forms.Form):
+	content = forms.CharField(widget=forms.Textarea, max_length=10)
```
EditForm với đủ các thành phần greeting_id, guestbook_name, content

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -2,7 +2,36 @@
+
+	def get_context_data(self, **kwargs):
+		self.__class__.greeting_id = long(self.request.GET.get('edited_message', 0))
+		self.__class__.guestbook_name = self.request.GET.get('guestbook_name', DEFAULT_GUESTBOOK_NAME)
```
không dùng biến của class,
tạo dict from request.GET, truyền vào Form thông qua constructor của Form  
https://docs.djangoproject.com/en/1.7/ref/forms/api/#dynamic-initial-values

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -2,7 +2,36 @@
+
+	def form_valid(self, form):
+		dict = {
+			'greeting_id': self.__class__.greeting_id,
```
không dùng biên của class,
lấy data từ `form.cleaned_data` ra dùng
https://docs.djangoproject.com/en/1.7/topics/forms/#the-view

asv-tungpt committed:
use form for receive and get infomation

phucnpm commented on djangoguestbook/guestbook/templates/guestbook/edit_page.html:
```diff
@@ -0,0 +1,17 @@
+</head>
+<body>
+	<div><input type="hidden" value={{ greeting_id }} name="greeting_id"></div>
+	<div><input type="hidden" value={{ guestbook_name }} name="guestbook_name"></div>
```
2 thẻ ẩn này có cần thiết ?

asv-tungpt committed:
Xoá thẻ

phucnpm commented on djangoguestbook/guestbook/forms.py:
```diff
@@ -4,3 +4,10 @@
+class EditForm(forms.Form):
+	content = forms.CharField(widget=forms.Textarea, max_length=10)
+	greeting_id = forms.IntegerField()
+	guestbook_name = forms.CharField()
```
greeting_id là không cho edit, hãy tìm cách ẩn đi, không hiển thị lúc edit

asv-tungpt committed:
Ẩn greeting_id field
